fuel-vm/src/interpreter/executors/main.rs

```diff
@@ -42,52 +44,113 @@ impl<T> Interpreter<PredicateStorage, T> {
     where
         Tx: ExecutableTransaction,
-        <Tx as IntoChecked>::Metadata: CheckedMetadata,
+        <Tx as IntoChecked>::CheckedMetadata: CheckedMetadata,
     {
         if !checked.transaction().check_predicate_owners(&params) {
             return Err(PredicateVerificationFailed::InvalidOwner);
         }
 
         let mut vm = Interpreter::with_storage(PredicateStorage::default(), params, gas_costs);
 
+        let mut cumulative_gas_used: Word = 0;
+
+        vm.init_predicate(checked.clone());
+
         // Needed for now because checked is only freed once the value is collected into a Vec
         #[allow(clippy::needless_collect)]
         let predicates: Vec<_> = (0..checked.transaction().inputs().len())
             .filter_map(|i| RuntimePredicate::from_tx(&params, checked.transaction(), i))
             .collect();
 
-        // Since we reuse the vm objects otherwise, we need to keep the actual gas here
-        let tx_gas_limit = checked.transaction().limit();
-        let mut remaining_gas = tx_gas_limit;
-
-        vm.init_predicate(checked);
-
         for predicate in predicates {
             // VM is cloned because the state should be reset for every predicate verification
             let mut vm = vm.clone();
 
-            vm.context = Context::Predicate { program: predicate };
-            vm.set_remaining_gas(remaining_gas);
+            let gas_used = predicate.gas_used();
+            vm.context = Context::PredicateVerification { program: predicate };
+            vm.set_gas(gas_used);
 
             if !matches!(vm.verify_predicate()?, ProgramState::Return(0x01)) {
                 return Err(PredicateVerificationFailed::False);
             }
 
-            remaining_gas = vm.registers[RegId::GGAS];
+            if vm.registers[RegId::GGAS] != 0 {
+                return Err(PredicateVerificationFailed::GasMismatch);
+            }
+            cumulative_gas_used = cumulative_gas_used
+                .checked_add(gas_used)
+                .expect("cumulative gas overflow");
         }
 
         Ok(PredicatesChecked {
-            gas_used: tx_gas_limit
-                .checked_sub(remaining_gas)
-                .ok_or_else(|| Bug::new(BugId::ID004, GlobalGasUnderflow))?,
+            gas_used: cumulative_gas_used,
         })
     }
+    /// Initialize the VM with the provided transaction and check all predicates defined in the
+    /// inputs.
+    ///
+    /// The storage provider is not used since contract opcodes are not allowed for predicates.
+    /// This way, its possible, for the sake of simplicity, it is possible to use
+    /// [unit](https://doc.rust-lang.org/core/primitive.unit.html) as storage provider.
+    ///
+    /// # Debug
+    ///
+    /// This is not a valid entrypoint for debug calls. It will only return a `bool`, and not the
+    /// VM state required to trace the execution steps.
+    pub fn estimate_predicates<Tx>(
```

Comment:
I would move this out of the `Interpreter` and make it a stadalone function as it doesn't take self anyway. Then it can have unit tests

Reply:
given that the check_predicates also does not operate on self, do you still feel this should be refactored out?

Reply:
Maybe not but it should at least have a unit test

Reply:
there is now a specific test in predicate.rs that addresses a transaction that fails predicate checking before estimation and succeeds after estimation